### ISSUE TITLE
Consistent-trees but in hdf5 format

### ISSUE THREAD

**Issue by manodeep** [ytree-project/ytree#46]
(Not really a bug-report, more of a question/feature request)

Thanks for continuing to develop ytree and congrats on the JOSS paper :)

We are hoping to use ytree for a project that uses Consistent-trees  but in an hdf5 format. Does ytree support such an use-case? 

Thanks,
Manodeep

**Comment by brittonsmith** [comment]
Not yet, but no reason why we couldn't. I didn't even know this was a thing. If you have some smallish (<= 100 MB) sample data, we should be able to do this no problem.

**Comment by manodeep** [comment]
It's not a thing (yet). We are planning to deliver some huge consistent-trees files and so we are converting the normal ascii files from CTrees to hdf5. 

Seems like adding the support would not be too onerous. Let me consult with the team, freeze the format, and then I will report back. Thanks again!

**Comment by brittonsmith** [comment]
As a work-free alternative, you could also try re-saving them in the ytree format with:
```
a = ytree.load(...)
a.save_arbor()
```

It may end up looking reasonably similar to the format you're working on.

**Comment by manodeep** [comment]
Thanks for that tip. Would ytree be able to load and convert treefiles in parallel? 

And will you please point me to the documentation for the arbor file format - I would like to evaluate the format...

**Comment by brittonsmith** [comment]
Those operations are not parallelized, but I think it would be very straightforward to do so, at least with mpi. I've been considering this for some time, but hadn't worked up the motivation.

Yes, documentation on the format... That would be a great thing to have. Here's a brief description, but I recommend taking some sample data and running the commands above to have a look.

The output format is a series of HDF5 files named (for a given name):
 - \<name>/\<name>.h5
 - \<name>/\<name>_0000.h5
 - \<name>/\<name>_0001.h5, etc

The first file contains metadata (cosmological parameters, box size, etc.) and the following information for each tree:
 - id of file where tree is located
 - start/end index within file for tree data
 - field values for the final (root) halo for each tree (mass, position, etc)

The numbered files 1D arrays of fields for multiple trees. Every tree is fully contained within a single file, i.e., no overlaps. The number of files is controlled by setting a maximum array size during the save process.

I'm open to making improvements to the format as well.

**Comment by manodeep** [comment]
Thanks for the details. Using Consistent-Trees lingo, can a forest (which comprises of multiple trees) be split across multiple files (i.e., could the constituent trees of the forest be located in different files)? 

Another question: Is the halo info stored as an array of structures or as a structure of arrays?

**Comment by brittonsmith** [comment]
At the moment, I only load a single tree_#_#_#.dat file (my data was never big enough to have more than one), so in that sense, yes, a forest could be split across multiple files. It should not be too difficult to add support for multiple .dat files and using the forests.list file.

Halo fields are stored as structures of arrays.

**Comment by brittonsmith** [comment]
One more quick update, I've got some rather large consistent-trees outputs to load in myself here, so I'm probably going to start looking at some optimizations for loading and saving.

**Comment by manodeep** [comment]
I am writing the CTrees->hdf5 (MPI-parallel) python converter - so you should be able to use that script out of the box. 

If it will be helpful, happy to have a chat on how to optimise. I have a couple of `C` repos that operate with Consistent-Trees formats -- [ConvertCtrees](https://github.com/manodeep/ConvertCTrees) for converting CTrees ascii into LHaloTree binary; and [parse_ctrees](https://github.com/manodeep/parse_ctrees) for a header-only generic reader for CTrees. Y. Y. Mao has a great utility for reading CTrees in python -- [SimulationAnalysis.py](https://bitbucket.org/yymao/helpers/src/master/helpers/SimulationAnalysis.py)

**Comment by manodeep** [comment]
@brittonsmith I have mostly finished up the Ctrees-ascii -> Ctrees->hdf5 format. The MPI-parallel code is in this [gist](https://gist.github.com/manodeep/1ce949bb51be1d93ba1ce374f556bc2d). The script is usable as is right now but I am planning to add a little more functionality in the near future. 

The basic idea for this new Ctrees-hdf5 format is that there is a container file `<forest>.h5`, with the actual forest data spread over files `<forest>_0.h5, <forest>_1.h5, ...`. These data files are embedded as external links into the container file, so you can loop over all datasets like so:

```python
# (un-tested code, to illustrate the format)
import h5py

fields = ['Mvir', 'Rvir', 'Vpeak']
with h5py.File('forest.h5', 'r') as hf:
    nfiles = hf['/'].attrs['Nfiles']
    for ifile in range(nfiles):
            forests = hf['File{}/Forests'.format(ifile)]
            # loop over any fields that you want to read
            for fld in fields:
                fld_vals = forests[fld][:]

```

I have assumed here (the default behaviour) where each halo property is written as a dataset (so as many datasets as the number of halo properties being written in the hdf5 file). The converter code also has the option to write out all properties of a halo contiguously. In that case, all of the properties will be stored in a single dataset `hf['File{}/halos'.format(ifile)]`. The code to create the header and setup the datasets is [here](https://gist.github.com/manodeep/1ce949bb51be1d93ba1ce374f556bc2d#file-convert_uchuu_ctrees_to_h5-py-L182-L234), and the code to create the container file is [here](https://gist.github.com/manodeep/1ce949bb51be1d93ba1ce374f556bc2d#file-convert_uchuu_ctrees_to_h5-py-L762-L777)

Based on feedback from Peter Behroozi, I am planning to add the five mergertree indices from `LHaloTree` convention -- `FirstHaloinFOFGroup, NextHaloinFOFGroup, FirstProgenitor, NextProgenitor, Descendant` into the hdf5 file as well. Here is the existing structure of the container file and one of the actual data files (ignore the filename as `tree_`, those filenames are now `forest_`)

[forest.dirtree.txt](https://github.com/ytree-project/ytree/files/4631663/tree.dirtree.txt)
[forest_0.dirtree.txt](https://github.com/ytree-project/ytree/files/4631664/tree_0.dirtree.txt)

How difficult will it be to add a reader for this CTrees-hdf5 format (including the mergertree indices) into ytree?

**Comment by brittonsmith** [comment]
@manodeep, sorry for taking so long to get back to you. Things are moving very slowly here lately.

Assuming all the same fields are present, I think this will not be much work at all to add a frontend for this format. I'm finally getting some time to work on this again, so I could probably do it relatively quickly. When things are ready, could you provide some sample data of no more than a couple hundred MB (~1 GB tops)?

**Comment by manodeep** [comment]
Hey no worries - we are in strange times now! 

I still have not managed to implement the mergertree indices, currently working on adding tests (and squashing bugs) for the initial conversion. I assume you would want a file with the mergertree indices included - yes?

**Comment by brittonsmith** [comment]
Thanks, strange times indeed. I'm happy to wait until the indices have been added, but I don't think I'll need them for the frontend. I've got enough to do at the moment, so entirely up to you.

**Comment by manodeep** [comment]
I have shared the test data-files through dropbox - you should have gotten an email. For starters, just use the `soa/forest.h5` file.

**Comment by manodeep** [comment]
Direct links to the files:

[soa/forest.h5](https://www.dropbox.com/s/e7x43mw94ku6knl/forest.h5?dl=1)
[soa/forest_0.h5](https://www.dropbox.com/s/22w6ghf7fw90ypp/forest_0.h5?dl=1)

**Comment by brittonsmith** [comment]
Thanks! Got them now.

**Comment by brittonsmith** [comment]
Ok, PR #53 has an initial implementation. In the PR, I outlined a couple issues/requests for things that could be added to the file format. Let me know what you think.

**Comment by manodeep** [comment]
I still haven't gotten a chance to implement the additional cosmological constants or the units. In the meantime, you can use the `metadata` field to parse those values directly

**Comment by brittonsmith** [comment]
> I still haven't gotten a chance to implement the additional cosmological constants or the units. In the meantime, you can use the `metadata` field to parse those values directly

I'm just getting around to this part, but the metadata attribute has everything we need and this will allow me to reuse the code we already have. I think we should just use this.

**Comment by manodeep** [comment]
Since the converter can also write out the forests as an array-of-structures, do you want to support that format as well? The difference is that there will be only one dataset in the aos case -- `Forests/halos` and all halo properties will be within that dataset. Here's the relevant setup code in the converter:

```python
    forests_grp = hf.create_group('Forests')
    if write_halo_props_cont:
        # Create a dataset for every halo property
        # For any given halo property, the value
        # for all halos will be written contiguously
        # (structure of arrays)
        for name, dtype in output_dtype.descr:
            forests_grp.create_dataset(name, (0,), dtype=dtype,
                                       chunks=chunks,
                                       compression=compression,
                                       maxshape=(None,))
    else:
        # Create a single dataset that contains all properties
        # of a given halo, then all properties of the next halo,
        # and so on (array of structures)
        forests_grp.create_dataset('halos', (0,),
                                   dtype=output_dtype,
                                   chunks=chunks,
                                   compression=compression,
                                   maxshape=(None,))

```

**Comment by brittonsmith** [comment]
I'm happy to support this format if you foresee it being used. If you have some sample data, I can add this. I don't know how this will work performance-wise within how ytree loads data.

**Comment by manodeep** [comment]
I don't know whether people will use that format but the converter can produce the output in the `SOA` (default, the format you just built support for) format as well as the `AOS` format. Here are two links for the `AOS` format:

[aos/forest.h5](https://www.dropbox.com/s/21bfo0l8jg64r0d/forest.h5?dl=0)
[aos/forest_0.h5](https://www.dropbox.com/s/adgtlcpft8k8p6e/forest_0.h5?dl=0)

**Comment by brittonsmith** [comment]
Resolved in PR #53!
